packages/data-context/src/data/ProjectConfigManager.ts

```diff
@@ -0,0 +1,790 @@
+    ])
+
+    this._cachedFullConfig = await this.buildBaseFullConfig(configFileContents, envFile, options, withBrowsers)
+
+    return this._cachedFullConfig
+  }
+
+  async getConfigFileContents () {
+    if (this._cachedLoadConfig?.initialConfig) {
+      return this._cachedLoadConfig?.initialConfig
+    }
+
+    return this.initializeConfig()
+  }
+
+  async loadCypressEnvFile () {
+    return this.#cypressEnv.loadCypressEnvFile()
+  }
+
+  async reloadCypressEnvFile () {
+    this.#cypressEnv = new CypressEnv({
+      envFilePath: this.#envFilePath,
+      validateConfigFile: (filePath, config) => {
+        this.validateConfigFile(filePath, config)
+      },
+      toLaunchpad: (...args) => {
+        this.options.toLaunchpad(...args)
+      },
+    })
+
+    return this.#cypressEnv.loadCypressEnvFile()
+  }
+
+  isTestingTypeConfigured (testingType: TestingType): boolean {
+    const config = this.#loadedConfigFile
+
+    if (!config) {
+      return false
+    }
+
+    if (!_.has(config, testingType)) {
+      return false
+    }
+
+    if (testingType === 'component') {
+      return Boolean(config.component?.devServer)
+    }
+
+    return true
+  }
+
+  private _pathToFile (file: string) {
+    return path.isAbsolute(file) ? file : path.join(this.options.projectRoot, file)
+  }
+
+  private initializeConfigWatchers () {
+    if (this.options.isRunMode) {
+      return
+    }
+
+    const configWatchers = this.addWatcher([
+      ...POTENTIAL_CONFIG_FILES.map((f) => this._pathToFile(f)),
+    ])
+
+    configWatchers.on('all', (event, file) => {
+      debug('WATCHER: config file event', event, file)
+
+      this.reloadConfig().catch(this.onLoadError)
+    })
+
+    configWatchers.on('error', (err) => {
+      debug('error watching config files %O', err)
+      this.options.onWarning(getError('UNEXPECTED_INTERNAL_ERROR', err))
+    })
+
+    const cypressEnvFileWatcher = this.addWatcher(this.#envFilePath)
+
+    cypressEnvFileWatcher.on('all', () => {
+      this.reloadConfig().catch(this.onLoadError)
+    })
+  }
+
+  private closeWatchers () {
+    for (const watcher of [...this._watchers.values(), ...Object.values(this._requireWatchers)]) {
+      // We don't care if there's an error while closing the watcher,
+      // the watch listener on our end is already removed synchronously by chokidar
+      watcher.close().catch((e) => {})
+    }
+    this._watchers = new Set()
+    this._requireWatchers = {}
+  }
+
+  destroy () {
+    if (this._eventsIpc) {
+      this._cleanupIpc(this._eventsIpc)
+    }
+
+    this.killChildProcesses()
+    this._state = 'pending'
+    this._cachedLoadConfig = undefined
+    this._cachedFullConfig = undefined
+    this._eventProcess = undefined
+    this._registeredEventsTarget = undefined
+    this.closeWatchers()
+  }
+}
+
+function pDeferFulfilled<T> (): pDefer.DeferredPromise<T> & {settled: boolean} {
+  const dfd = pDefer<T>()
+  let settled = false
+  const { promise, resolve, reject } = dfd
+
```

Comment:
From when I was looking into this previously, I think once we rework things a bit more, we should be able to get rid of the need for this.